ga4gh/backends.py

```diff
@@ -309,6 +310,109 @@ def getMetadata(self):
         return ret
 
 
+def _encode_value(value):
+    if isinstance(value, (list, tuple)):
+        return [ str(v) for v in value ]
+    else:
+        return [str(value)]
+
+
+class HtslibVariantSet(VariantSet):
+    """
+    Class representing a single variant set backed by a tabix directory.
+    """
+    def __init__(self, variantSetId, vcfPath, use_bcf=True):
+        self._variantSetId = variantSetId
+        self._created = protocol.convertDatetime(datetime.datetime.now())
+        self._chromFileMap = {}
+
+        for pattern in ['*.bcf', '*.vcf.gz']:
+            for vcfFile in glob.glob(os.path.join(vcfPath, pattern)):
+                self._addFile(vcfFile)
+
+    def _addFile(self, vcfFile):
+        varFile = pysam.BCFFile(vcfFile)
+        if not varFile.index:
+            raise Exception("VCF/BCF files must be indexed")
+        for chrom in varFile.index:
+            has_records = bool(next(varFile.fetch(chrom), None))
```

Comment:
I agree the implementation is less than crystal clear for those not deeply
rooted in the nuances of Python's iterator API.  I'm not conceptually
against adding such a method, but it seems a very uncommon use case that is
fairly specific to attempting to treat a collection of files as a
"dataset".  How about a helper function:

_nothing = object()
def is_empty_iter(iter):

    '''returns True iff iter is empty or exhausted'''

    return next(iter, _nothing) is _nothing


and

    has_records = is_empty_iter(varFile.fetch(chrom))


On Wed, Jan 14, 2015 at 8:37 AM, Jerome Kelleher <notifications@github.com>
wrote:

> In ga4gh/backends.py
> <https://github.com/jeromekelleher/server/pull/1#discussion-diff-22949144>
> :
>
> > +    """
> > +    def __init__(self, variantSetId, vcfPath, use_bcf=True):
> > +        self._variantSetId = variantSetId
> > +        self._created = protocol.convertDatetime(datetime.datetime.now())
> > +        self._chromFileMap = {}
> > +
> > +        for pattern in ['*.bcf', '*.vcf.gz']:
> > +            for vcfFile in glob.glob(os.path.join(vcfPath, pattern)):
> > +                self._addFile(vcfFile)
> > +
> > +    def _addFile(self, vcfFile):
> > +        varFile = pysam.BCFFile(vcfFile)
> > +        if not varFile.index:
> > +            raise Exception("VCF/BCF files must be indexed")
> > +        for chrom in varFile.index:
> > +            has_records = bool(next(varFile.fetch(chrom), None))
>
> This seems a little clumsy - could we get an API call for
> varFile.has_records() instead please?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jeromekelleher/server/pull/1/files#r22949144>.
>